contrib/ruby/ext/trilogy-ruby/cext.c

```diff
@@ -211,10 +253,23 @@ static int _cb_ruby_wait(trilogy_sock_t *sock, trilogy_wait_t wait)
     }
 
     int fd = trilogy_sock_fd(sock);
+
+#ifdef TRILOGY_RB_IO_WAIT
+    if (ctx->io == Qnil) {
+        ctx->io = rb_io_fdopen(fd, O_RDWR, NULL);
```

Comment:
`trilogy_data_type` is marked with `RUBY_TYPED_WB_PROTECTED`, so you have to trigger a write barrier here.

Problem is you don't have access to the `VALUE` here.

Reply:
Okay, so we should pass around the `VALUE` then?

Reply:
Either that, or create the IO eagerly in a place where you have access to it.